Add support for clearing the count repetitions by pressing Esc

yarbelk opened the pull request:
Moved the check for pressing Esc to before where the command event keys are processed - and it will clear out the number of repetitions entered.

This was going to be part of implementing F and f as commands: Both of which require you to be able to enter addtional characters after the command.  

First pass didn't work so well, so here is just the escape part.

yarbelk committed:
Add support for clearing the count repetitions by pressing Esc

kisielk commented on mode/normal.go:
```diff
@@ -44,6 +45,18 @@ func (m *normalMode) OnKey(ev *termbox.Event) {
+	switch ev.Ch {
+	case 0x0:
+		switch ev.Key {
+		case termbox.KeyEsc:
```
Let's put this in `func (*m normalMode) Reset()` and also call that at L231

yarbelk committed:
Move the command reset code to m.Reset

- It is used in multiple places, so refactor to new function on
  normalMode.

yarbelk committed:
OnKey using channels and EditorMode as an interface

- need some way of dealing with streams of commands comming in that lets
  us assemble composite commands - seperating out the implementation of
  parsing what a event results in helps going that way
- Reset then needs to be used across multiple modes, but it means
  something different in each - so we introduce the EditorMode interface

- NOTE: broken atm

yarbelk commented:
Did so, but let me make more changes before pulling in - I want to make it part of command

yarbelk committed:
WIP implement editor.Mode across every mode

yarbelk committed:
remove newMode test variable from code

yarbelk committed:
export all Modes

yarbelk committed:
return pointers from all NewMode functions

yarbelk committed:
revert go routine version of OnKey

yarbelk committed:
Run reset command if esc is pressed

yarbelk commented:
ready @kisielk

kisielk commented on mode/insert.go:
```diff
@@ -6,22 +6,35 @@ import (
 }
 
-func (m insertMode) Enter(editor *editor.Editor) {
+func NewInsertMode(editor *editor.Editor, count int) (m *InsertMode) {
```
I prefer to avoid named return variables unless necessary for clarity

kisielk commented:
Thanks, I think it's looking good, but I think we may still be conflating the usage of "Reset" and "Exit" in some places. I think we should probably use "Reset" to mean reset the state of the current mode and stay there, such as when you push esc during a key sequence in normal mode. "Exit" should return you to the previous mode. What do you think?

yarbelk commented:
I think you're on to something.  the question is confused by the use of the 'Esc' key.  Sometimes its one, sometimes the other.

example:

`Esc` in normal mode with digitts typed in -> Reset
`Esc` in command mode -> Escape.

I will do a pass through and see if i'm crossing these and clean up your comments

yarbelk committed:
In CommandMode, switch implememntation of Reset and Exit

 - they were backwards
 - Add some comments to Exit, Reset (generally these seem to be NOOP)
 - Note - there are known bugs in this that are fixed in other pull
   requests

yarbelk committed:
Remove named return variable in Enter
